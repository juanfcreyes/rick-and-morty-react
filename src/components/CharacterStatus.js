export const CharacterStatus = ({ status }) => {
    if (status === "Alive") {
		return <span className="badge bg-success">{status}</span>;
	} else if (status === "Dead") {
		return <span className="badge bg-danger">{status}</span>;
	}
	return <div></div>;
};
