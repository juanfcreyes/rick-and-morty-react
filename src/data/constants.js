export const IMAGE_URL = "https://rickandmortyapi.com/api/character/avatar/";
export const API_URL = "https://rickandmortyapi.com/api/character/";