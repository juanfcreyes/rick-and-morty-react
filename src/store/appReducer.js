import { createAsyncThunk, createSlice, createSelector } from "@reduxjs/toolkit";
import { API_URL } from "../data/constants";

const initialState = {
	active: null,
	data: [],
	status: "idle",
	error: null,
	next: null,
	previous: null,
	pokemonsDetails: {},
};

export const callApi = createAsyncThunk("app/callApi", async (url = API_URL) => {
    const response = await fetch(url);
	const body = await response.json();
	return body;
});

const appSlider = createSlice({
	initialState,
	name: "app",
	reducers: {
	},
	extraReducers: {
		[callApi.pending]: (state) => {
			state.status = "loading";
		},
		[callApi.fulfilled]: (state, action) => {
			state.status = "succeeded";
			const { results, info } = action.payload;
            console.log(action.payload);
			state.data = results.map((character) => ({
                id: character.id,
                name: character.name,
                status: character.status,
                gender: character.gender,
                type: character.type,
                origin: character.origin.name || ""
            }));
			state.next = info.next;
			state.previous = info.prev;
		},
		[callApi.rejected]: (state, action) => {
			state.status = "failed";
			state.error = action.error.message;
		}
	},
});

export const { setActive } = appSlider.actions;

const selectApp = (state) => state.app;
export const selectActive = createSelector(selectApp, (app) => app.active);

export const selectById = createSelector(
	[selectApp, (_, id) => id],
	(app, id) => app.data.find(character => character.id === id)
);

export const appReducer = appSlider.reducer;
