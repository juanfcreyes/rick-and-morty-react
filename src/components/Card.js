import { useSelector } from "react-redux";
import { IMAGE_URL } from "../data/constants";
import { selectById } from "../store/appReducer";
import { CharacterStatus } from "./CharacterStatus";

export const Card = ({ id }) => {
	const character = useSelector((state) => selectById(state, id));

	return (
		<div className='card mb-4 '>
			<div className='row'>
				<div className='col-4'>
					<img
						width='130px'
                        height='200px'
						src={`${IMAGE_URL}${character.id}.jpeg`}
						alt={character.name}
					></img>
				</div>
				<div className='col-8'>
					<p>
						<strong>{character.name}</strong>
					</p>
					<p>
						<strong> Gender: </strong> {character.gender}
					</p>

					<p>
						<strong> Origin: </strong> {character.origin}
					</p>
					<div className='row'>
						<div className='col-6 '>
							<strong> Type: </strong> {character.type}
						</div>
						<div className='col-6'>
							<CharacterStatus status={character.status}></CharacterStatus>
						</div>
					</div>
				</div>
			</div>
		</div>
	);
};
