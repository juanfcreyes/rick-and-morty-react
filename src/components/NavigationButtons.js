import { useDispatch, useSelector } from "react-redux";
import { callApi } from "../store/appReducer";

export const NavigationButtons = () => {

    const { next, previous } = useSelector((state) => state.app);
    
    const dispatch = useDispatch();
	
	const handleOnNext = () => {
		dispatch(callApi(next));
	};

	const handleOnPrevious = () => {
		dispatch(callApi(previous));
	};

    return (
		<div className='container mt-4'>
			<div className='row justify-content-evenly'>
				<div className='col-6'>
					<button
						onClick={handleOnPrevious}
						disabled={!previous}
						className='btn btn-dark'>
						Previuos
					</button>
				</div>
				<div className='col-6'>
					<button onClick={handleOnNext} disabled={!next} className='btn btn-dark float-end'>
						Next
					</button>
				</div>
			</div>
		</div>
	);
};
