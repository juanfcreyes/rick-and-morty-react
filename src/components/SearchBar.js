import { useState } from "react";
import { useDispatch } from "react-redux";
import { API_URL } from "../data/constants";
import { callApi } from "../store/appReducer";

export const SearchBar = () => {
    const dispatch = useDispatch();

	const [query, setQuery] = useState("");

	const handleOnChange = (event) => {
		setQuery(event.target.value);
	};

	const hanldeOnSubmit = (event) => {
		event.preventDefault();
        if (query) {
            dispatch(callApi(`${API_URL}?name=${query}`))
        } else {
            dispatch(callApi())
        }
	};

	return (
		<form className='row' onSubmit={hanldeOnSubmit}>
			<div className='col-10'>
				<input
					type='text'
					className='form-control'
					id='query'
					placeholder='Insert a character name'
					onChange={handleOnChange}
				/>
			</div>
			<div className='col-2'>
				<button type='submit' className='btn btn-dark mb-3'>
					Search
				</button>
			</div>
		</form>
	);
};
