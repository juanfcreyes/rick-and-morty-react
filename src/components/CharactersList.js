import { useMemo } from "react";
import { useSelector } from "react-redux";
import { Card } from "./Card";

export const CharacterList = () => {
	const { data: characters } = useSelector((state) => state.app);

    return useMemo(
		() => (
			<div className='row list-container'>
				{characters.map((character) => (
					<div className='col-4' key={character.id}>
						{<Card key={character.id} id={character.id}></Card>}
					</div>
				))}
			</div>
		),
		[characters],
	);
};
