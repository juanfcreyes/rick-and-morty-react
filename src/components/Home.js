import { useDispatch } from "react-redux";
import { callApi } from "../store/appReducer";
import { CharacterList } from "./CharactersList";
import { useEffect } from "react";
import { NavigationButtons } from "./NavigationButtons";
import { SearchBar } from "./SearchBar";

export const Home = () => {

	const dispatch = useDispatch();
    
    useEffect(() => {
		dispatch(callApi());
	}, [dispatch]);

	return (
		<div>
			<div className='row'>
				<div className='col-12'>
					<SearchBar></SearchBar>
					<CharacterList></CharacterList>
                    <NavigationButtons></NavigationButtons>
				</div>
			</div>
		</div>
	);
};
