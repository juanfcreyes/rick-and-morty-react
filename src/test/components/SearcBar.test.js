import { mount } from "enzyme/build";
import { Provider } from "react-redux";
import { SearchBar } from "../../components/SearchBar";
import '@testing-library/jest-dom';
import configureStore from 'redux-mock-store';
import { callApi } from "../../store/appReducer";
import { API_URL } from "../../data/constants";

const mocktore = configureStore();
const initialState = {};
const store = mocktore(initialState);
store.dispatch = jest.fn();

jest.mock("../../store/appReducer", () => ({
    callApi: jest.fn(),
}));

describe('Test over SearchBar', () => { 

    const wrapper = mount(
    <Provider store={store}>
        <SearchBar></SearchBar>
    </Provider>
    );

    test('should show correctly ', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should dispath callApi with the default url ', () => {
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        expect(callApi).toHaveBeenCalled();
        expect(callApi).toHaveBeenCalledWith();
    });
    
    test('should dispath callApi with the default query url ', () => {
        wrapper.find('input').simulate('change', {
            target: {
                value: 'rick'
            }
        });    
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        expect(callApi).toHaveBeenCalled();
        expect(callApi).toHaveBeenCalledWith(`${API_URL}?name=rick`);
    });    
});