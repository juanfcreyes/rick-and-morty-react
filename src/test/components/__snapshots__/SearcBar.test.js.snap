// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test over SearchBar should show correctly  1`] = `
<form
  className="row"
  onSubmit={[Function]}
>
  <div
    className="col-10"
  >
    <input
      className="form-control"
      id="query"
      onChange={[Function]}
      placeholder="Insert a character name"
      type="text"
    />
  </div>
  <div
    className="col-2"
  >
    <button
      className="btn btn-dark mb-3"
      type="submit"
    >
      Search
    </button>
  </div>
</form>
`;
